src/mon/OSDMonitor.cc

```diff
@@ -1625,18 +1625,21 @@ bool OSDMonitor::can_mark_in(int i)
   return true;
 }
 
-void OSDMonitor::check_failures(utime_t now)
+bool OSDMonitor::check_failures(utime_t now)
 {
+  bool propose_now = false;
   for (map<int,failure_info_t>::iterator p = failure_info.begin();
        p != failure_info.end();
        ++p) {
     if (can_mark_down(p->first)) {
-      check_failure(now, p->first, p->second);
+      check_failure(now, p->first, p->second, propose_now ? NULL : &propose_now);
     }
   }
+  return propose_now;
```

Comment:
i agree with @xiaoxichen , the propose_timer will be cancelled if the `tick()` decides to propose the pending changes at seeing a failed OSD, even if a pending change for the failed OSD was scheduled by the `PaxosService::prepare_update()` because of a failure report.